voucherify/client.py

```diff
@@ -0,0 +1,166 @@
+
+        if result and isinstance(result, dict) and result.get('error'):
+            raise VoucherifyError(result)
+
+        return result
+
+    def list(self, query):
+        path = '/vouchers/'
+
+        return self.request(
+            path,
+            query
+        )
+
+    def get(self, code):
+        path = join_url('/vouchers/', quote(code))
+
+        return self.request(
+            path
+        )
+
+    def create(self, voucher):
+        code = voucher.get('code', '')
+        path = join_url('/vouchers/', quote(code))
+
+        return self.request(
+            path,
+            data=json.dumps(voucher),
+            method='POST'
+        )
+
+    def enable(self, code):
+        path = join_url('/vouchers/', quote(code), '/enable')
+
+        return self.request(
+            path,
+            method='POST'
+        )
+
+    def disable(self, code):
+        path = join_url('/vouchers/', quote(code), '/disable')
+
+        return self.request(
+            path,
+            method='POST'
+        )
+
+    def redemption(self, code):
+        path = join_url('/vouchers/', quote(code), '/redemption')
+
+        return self.request(
+            path
+        )
+
+    def redemptions(self, query):
+        path = '/redemptions'
+
+        return self.request(
+            path,
+            args=query
+        )
+
+    def redeem(self, code, tracking_id=None):
+        voucher = {}
+
+        if code and isinstance(code, dict) and code.get('error'):
+            voucher = code
+            code = voucher['code']
+            del voucher['code']
+
+        path = join_url('/vouchers/', quote(code), '/redemption')
+
+        if tracking_id:
+            join_url_params(path, {'tracking_id': tracking_id})
+
+        return self.request(
+            path,
+            method='POST',
+            data=json.dumps(voucher),
+        )
+
+    def rollback(self, redemption_id, reason=None):
+        path = join_url('/redemptions/', redemption_id, '/rollback')
+
+        if reason:
+            join_url_params(path, {'reason': reason})
+
+        return self.request(
+            path,
+            method='POST'
+        )
+
+    def publish(self, campaign_name):
+        path = join_url_params('/vouchers/publish', {'campaign': campaign_name})
+
+        return self.request(
+            path,
+            method='POST'
+        )
+
+
+class VoucherifyError(Exception):
+    def __init__(self, result):
+        self.result = result
+        self.code = None
+        self.message = None
+
+        try:
+            self.type = result['error_code']
+            self.message = result['error_msg']
+        except:
+            self.type = ''
+            self.message = result
+
+        Exception.__init__(self, self.message)
+
+
+def join_url(url, *paths):
+    for path in paths:
+        url = re.sub(r'/?$', re.sub(r'^/?', '/', path), url)
+    return url
```

Comment:
Why so complex here? we don't need regex at all imho.
Why not just simply concat the strings in URLs? e.g. `path = '/vouchers/' + quote(code) + '/redemption'`